src/applicationManager.ts

```diff
@@ -528,3 +546,239 @@ async function getGraphAPIRoles(token: string, graphApiPermissions: string[]) {
   }
   return { appRoleIds, graphAPIObjectId };
 }
+
+/**
+ * Updates an Azure Entra Application, with a collection of App Roles
+ * @param applicationId The object ID of application registration
+ * @param appRoles appRoles to update the application with
+ */
+export async function updateApplicationAppRoles({
+  token,
+  applicationId,
+  appRoles,
+}: {
+  token: string;
+  applicationId: string;
+  appRoles: AppRole[];
+}): Promise<void> {
+  let body = {
+    appRoles: appRoles,
+  };
+  const result = await fetch(
+    `https://graph.microsoft.com/v1.0/applications/${applicationId}`,
+    {
+      method: "PATCH",
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "Content-Type": "application/json",
+      },
+      body: JSON.stringify(body),
+    },
+  );
+  await errorHandler("updating application config", result);
+}
+
+/**
+ * Necessary where an app role is deleted, they must all be disabled first. Easier to disable by default and enable when updating.
+ */
+async function disableAppRoles(
+  applicationId: string,
+  token: string,
+  appRoles: AppRoles,
+) {
+  const app = await readApplication({ token, applicationId });
+  let appRolesJson: Array<AppRole> = app.appRoles;
+
+  // Finds pairs of displayName and ID of app roles to compare existing app roles with incoming app roles
+  const incomingAppRolePairs = appRoles.map(
+    ({ displayName, id }) => `name:${displayName}_id:${id}`,
+  );
+  const existingAppRolePairs = appRolesJson.map(
+    ({ displayName, id }) => `name:${displayName}_id:${id}`,
+  );
+  const commonPairs = existingAppRolePairs.filter((pair) =>
+    incomingAppRolePairs.includes(pair),
+  );
+
+  // Break here if app roles are the same as current file - not disabling all roles if unnecessary
+  if (!(commonPairs.length < existingAppRolePairs.length)) {
```

Comment:
this code is a bit confusing, I think it would be easier to read if it was called something like: remainingAppRoles

then this if statement would be:

```suggestion
  if (remainingAppRoles.length === 0 && incomingAppRolePairs.length > 0) {
```

Reply:
`commonPairs` is a collection of roles that already exist **and** are present in the yaml file (based on display names and ids matching) --- if this list of common pairs is shorter than what's trying to be created, a role must've been deleted / not present in yaml file

I'm not sure the suggestion would achieve the same thing, returning if there are no matching pairs (sometimes only one role may be deleted etc), seems to be opposite to what I was going for?

Reply:
the code is very confusing to me.

mostly the if statement but i think renaming the variable helps too

Reply:
The if statement is just there to break before disabling the roles if all the app roles in the yaml file match what's in Azure - will see if I can rename them to make it a bit clearer though

Reply:
https://github.com/hmcts/azure-app-proxy-manager/pull/122/commits/22685e97905646c4130692d294f8a6dffc98ee4d - hopefully this makes some more sense

I need commonPairs so that it doesn't just to length based matching - but checks details too